lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs

```diff
@@ -0,0 +1,172 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+-- Copyright: © 2022 IOHK
+-- License: Apache-2.0
+--
+-- Computing minimum UTxO values.
+--
+module Cardano.Wallet.Shelley.MinimumUTxO
+    ( computeMinimumCoinForUTxO
+    , maxLengthCoin
+    , maxLengthAddress
+    , unsafeLovelaceToWalletCoin
+    , unsafeValueToLovelace
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..) )
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Primitive.Types.MinimumUTxO
+    ( MinimumUTxO (..), MinimumUTxOForShelleyBasedEra (..) )
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle (..) )
+import Cardano.Wallet.Primitive.Types.TokenMap
+    ( TokenMap )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( TxOut (..) )
+import Cardano.Wallet.Shelley.Compatibility
+    ( toCardanoTxOut )
+import Data.Function
+    ( (&) )
+import Data.IntCast
+    ( intCast, intCastMaybe )
+import Data.Word
+    ( Word64, Word8 )
+import GHC.Stack
+    ( HasCallStack )
+import Numeric.Natural
+    ( Natural )
+
+import qualified Cardano.Api.Shelley as Cardano
+import qualified Data.ByteString as BS
+
+-- | Computes a minimum 'Coin' value for a 'TokenMap' that is destined for
+--   inclusion in a transaction output.
+--
+computeMinimumCoinForUTxO :: HasCallStack => MinimumUTxO -> TokenMap -> Coin
+computeMinimumCoinForUTxO = \case
+    MinimumUTxONone ->
+        const (Coin 0)
+    MinimumUTxOConstant c ->
+        const c
+    MinimumUTxOForShelleyBasedEraOf pp ->
+        computeMinimumCoinForShelleyBasedEra pp
+
+computeMinimumCoinForShelleyBasedEra
+    :: HasCallStack
+    => MinimumUTxOForShelleyBasedEra
+    -> TokenMap
+    -> Coin
+computeMinimumCoinForShelleyBasedEra
+    (MinimumUTxOForShelleyBasedEra era pp) tokenMap =
+        extractResult $
+        Cardano.calculateMinimumUTxO era
+            (embedTokenMapWithinPaddedTxOut era tokenMap)
+            (Cardano.fromLedgerPParams era pp)
+  where
+    extractResult :: Either Cardano.MinimumUTxOError Cardano.Value -> Coin
+    extractResult = \case
+        Right value ->
+            -- We assume that the returned value is a non-negative ada quantity
+            -- with no other assets. If this assumption is violated, we have no
+            -- way to continue, and must raise an error:
+            value
+                & unsafeValueToLovelace
+                & unsafeLovelaceToWalletCoin
+        Left e ->
+            -- We assume that the provided protocol parameters record has all
+            -- the required parameters for the given era. If this assumption is
+            -- violated, we have no way to continue, and must raise an error:
```

Comment:
Suggestion: explain why this should be impossible. (The `MinimumUTxOForShelleyBasedEra` constructor already requires an era-specific protocol parameters object, which should mean that missing fields are impossible.)

Reply:
Fixed in a2c414ad3e44d1af900280b9a890bb398783e400.